launcher/JavaDownloader.cpp

```diff
@@ -0,0 +1,321 @@
+        if (type == "directory") {
+            FS::ensureFolderPathExists(file);
+        } else if (type == "link") {
+            // this is linux only !
+            auto target = FS::PathCombine(file, "../" + Json::requireString(Json::requireObject(list, paths), "target"));
+            QFile(target).link(file);
+        } else if (type == "file") {
+            // TODO download compressed version if it exists ?
+            auto raw = Json::requireObject(Json::requireObject(Json::requireObject(list, paths), "downloads"), "raw");
+            auto isExec = Json::ensureBoolean(Json::requireObject(list, paths), "executable", false);
+            auto f = File{ file, Json::requireString(raw, "url"), QByteArray::fromHex(Json::ensureString(raw, "sha1").toLatin1()), isExec };
+            toDownload.push_back(f);
+        }
+    }
+    auto elementDownload = new NetJob("JRE::FileDownload", APPLICATION->network());
+    for (const auto& file : toDownload) {
+        auto dl = Net::Download::makeFile(file.url, file.path);
+        dl->addValidator(new Net::ChecksumValidator(QCryptographicHash::Sha1, file.hash));
+        if (file.isExec) {
+            connect(dl.get(), &Net::Download::succeeded,
+                    [file] { QFile(file.path).setPermissions(QFile(file.path).permissions() | QFileDevice::Permissions(0x1111)); });
+        }
+        elementDownload->addNetAction(dl);
+    }
+    connect(elementDownload, &NetJob::finished, [elementDownload, this] {
+        disconnect(this, &Task::aborted, elementDownload, &NetJob::abort);
+        elementDownload->deleteLater();
+    });
+    connect(elementDownload, &NetJob::progress, this, &JavaDownloader::progress);
+    connect(elementDownload, &NetJob::failed, this, &JavaDownloader::emitFailed);
+
+    connect(this, &Task::aborted, elementDownload, &NetJob::abort);
+    connect(elementDownload, &NetJob::succeeded, [this] { emitSucceeded(); });
+    elementDownload->start();
+}
+void JavaDownloader::downloadAzulMeta(const QString& OS, bool isLegacy, const NetJob* netJob)
+{
+    setStatus(tr("Querying Azul meta"));
+    QString javaVersion = isLegacy ? QString("8.0") : QString("17.0");
+
+    QString azulOS;
+    QString arch;
+    QString bitness;
+
+    mojangOStoAzul(OS, azulOS, arch, bitness);
+    auto metaResponse = new QByteArray();
+    auto downloadJob = new NetJob(QString("JRE::QueryAzulMeta"), APPLICATION->network());
+    downloadJob->addNetAction(
+        Net::Download::makeByteArray(QString("https://api.azul.com/zulu/download/community/v1.0/bundles/?"
+                                             "java_version=%1"
+                                             "&os=%2"
+                                             "&arch=%3"
+                                             "&hw_bitness=%4"
+                                             "&ext=zip"          // as a zip for all os, even linux NOTE !! Linux ARM is .deb only !!
```

Comment:
maybe consider removing this comment, since it doesn't seem to be true. I see aarch32hf.tar.gz and aarch64.tar.gz files for multiple java versions for linux

```
  {
    "id": 34024, 
    "java_version": [
      17, 
      0, 
      4
    ], 
    "jdk_version": [
      17, 
      0, 
      4, 
      8
    ], 
    "name": "zulu17.36.15-ca-jdk17.0.4-linux_aarch32hf.tar.gz", 
    "openjdk_build_number": 8, 
    "url": "https://cdn.azul.com/zulu-embedded/bin/zulu17.36.15-ca-jdk17.0.4-linux_aarch32hf.tar.gz", 
    "zulu_version": [
      17, 
      36, 
      15, 
      0
    ]
  }, 
```
```
  {
    "id": 34985, 
    "java_version": [
      17, 
      0, 
      4, 
      1
    ], 
    "jdk_version": [
      17, 
      0, 
      4, 
      1
    ], 
    "name": "zulu17.36.17-ca-jre17.0.4.1-linux_aarch64.tar.gz", 
    "openjdk_build_number": 1, 
    "url": "https://cdn.azul.com/zulu/bin/zulu17.36.17-ca-jre17.0.4.1-linux_aarch64.tar.gz", 
    "zulu_version": [
      17, 
      36, 
      17, 
      0
    ]
  }, 
```

the urls for this are under zulu-embedded for arm32 so I'm not sure if they are missing files compared to the standard versions